domain/application/service/containerimageresourcestore.go

```diff
@@ -4,26 +4,91 @@
 package service
 
 import (
+	"bytes"
 	"context"
+	"encoding/json"
 	"io"
 
-	"github.com/juju/juju/core/errors"
+	coreresource "github.com/juju/juju/core/resources"
+	"github.com/juju/juju/domain/application/resource"
+	charmresource "github.com/juju/juju/internal/charm/resource"
+	"github.com/juju/juju/internal/docker"
+	"github.com/juju/juju/internal/errors"
 )
 
-// ContainerImageResourceState provides methods for interacting
-// with the container image resource store.
-type ContainerImageResourceState interface {
-}
-
-func newContainerImageResourceStore(st ContainerImageResourceState) *containerImageResourceStore {
+func newContainerImageResourceStore(st ContainerImageMetadataState) *containerImageResourceStore {
 	return &containerImageResourceStore{st: st}
 }
 
 // containerImageResourceStore is a ResourceStore for container image resource types.
 type containerImageResourceStore struct {
-	st ContainerImageResourceState
+	st ContainerImageMetadataState
+}
+
+// Get returns an io.ReadCloser for a resource in the resource store.
+func (s containerImageResourceStore) Get(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+) (r io.ReadCloser, size int64, err error) {
+	metadata, err := s.st.Get(ctx, resourceUUID)
+	if err != nil {
+		return nil, 0, errors.Errorf("getting container image metadata from state: %w", err)
+	}
+
+	details := docker.DockerImageDetails{
+		RegistryPath: metadata.RegistryPath,
+		ImageRepoDetails: docker.ImageRepoDetails{
+			BasicAuthConfig: docker.BasicAuthConfig{
+				Username: metadata.Username,
+				Password: metadata.Password,
+			},
+		},
+	}
+	data, err := json.Marshal(details)
+	if err != nil {
+		return nil, 0, errors.Errorf("marshelling container image metadata: %w", err)
+	}
+	infoReader := bytes.NewReader(data)
+	length := infoReader.Len()
+	return io.NopCloser(infoReader), int64(length), nil
+}
+
+// Put stores data from io.Reader in the resource store at the
+// path specified in the resource.
+func (s containerImageResourceStore) Put(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+	r io.Reader,
+	size int64,
+	fingerprint charmresource.Fingerprint,
+) (resource.ResourceStorageUUID, error) {
+	respBuf := new(bytes.Buffer)
+	_, err := respBuf.ReadFrom(r)
+	if err != nil {
+		return nil, errors.Errorf("reading container image resource: %w", err)
+	}
+
+	dockerDetails, err := docker.UnmarshalDockerResource(respBuf.Bytes())
+	if err != nil {
+		return nil, errors.Errorf("unmarsheling container image metadata: %w", err)
```

Comment:
**todo:** Unless it is again some british weirdness, i think it is `unmarshalling` right ?